### PR TITLE
Pkcs11 log update

### PR DESCRIPTION
**Pull request by lundinc2**
Pkcs11 log update

Description
-----------
Use C-SDK Style log implementation for PKCS #11.

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [x] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: projects/cypress/CY8CKIT_064S0S2_4343W/mtb/jitp/config_files/iot_pkcs11_config.h [hunk 1]
```diff
@@ -32,8 +32,16 @@
 #ifndef _AWS_PKCS11_CONFIG_H_
 #define _AWS_PKCS11_CONFIG_H_
 
-/* A non-standard version of C_INITIALIZE should be used by this port. */
-/* #define pkcs11configC_INITIALIZE_ALT */
+/* Logging configuration for the PKCS #11 library. */
+#ifndef LIBRARY_LOG_NAME
```
Comment (aggarw13) [inline projects/cypress/CY8CKIT_064S0S2_4343W/mtb/jitp/config_files/iot_pkcs11_config.h:36]: Would be good to have the comment section similar to what [we have on C-SDK](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/development/demos/mqtt/mqtt_demo_basic_tls/mqtt_config.h#L24-L34) that describes how these macros tie the PKCS library with the logging stack.

You also need to include `logging_levels.h` before the definition of these macros for accessing the `LOG_XX` macros

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 2]
```diff
@@ -556,19 +553,25 @@ static CK_RV prvMbedTLS_Initialize( void )
...
+                                                mbedtls_entropy_func,
+                                                &xP11Context.xMbedEntropyContext,
+                                                NULL,
+                                                0 );
+
+        if( 0 != lMbedTLSResult )
         {
+            LogError( ( "Failed to seed the mbed TLS DRBG for the PKCS #11 module with error %s",
+                        mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ) ) );
+            LogError( ( "%s", mbedtlsLowLevelCodeOrDefault( lMbedTLSResult ) ) );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:566]: The high and low level code strings can be logged in the same call
```suggestion
            LogError( ( "Failed to seed the mbed TLS DRBG for the PKCS #11 module: mbedTLSError=%s : %s",
                        mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ),
                        mbedtlsLowLevelCodeOrDefault( lMbedTLSResult ) ) );
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 3]
```diff
@@ -679,10 +683,12 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
                                 const mbedtls_pk_context * pxMbedContext )
 {
     CK_RV xResult = CKR_OK;
-    int32_t lMbedReturn = 0;
+    int32_t lMbedTLSResult = 0;
     CK_BBOOL xBool;
     mbedtls_rsa_context * pxRsaContext = ( mbedtls_rsa_context * ) pxMbedContext->pk_ctx;
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:688]: `configASSERT` for `pxMbedContext` before de-referencing it?

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c (hunk 2 above)
Comment (aggarw13) [reply libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:566]: Same for other logging sites of mbedTLS error strings

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 4]
```diff
@@ -790,10 +796,12 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
         /* See explanation in prvCheckValidSessionAndModule for this exception. */
         /* coverity[misra_c_2012_rule_10_5_violation] */
         CK_BBOOL xBool = ( CK_BBOOL ) CK_FALSE;
-        int32_t lMbedReturn = 0;
+        int32_t lMbedTLSResult = 0;
         CK_RV xResult = CKR_OK;
         mbedtls_ecp_keypair * pxKeyPair = ( mbedtls_ecp_keypair * ) pxMbedContext->pk_ctx;
 
+        configASSERT( pxKeyPair != NULL );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:803]: `configASSERT( pxMbedContext != NULL)` before de-referencing it on line 801?

File: projects/cypress/CY8CKIT_064S0S2_4343W/mtb/jitp/config_files/iot_pkcs11_config.h (hunk 1 above)
Comment (lundinc2) [reply projects/cypress/CY8CKIT_064S0S2_4343W/mtb/jitp/config_files/iot_pkcs11_config.h:36]: Added this comment. Thanks for calling out `logging_levels.h` missed copying that into most of thes

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c (hunk 3 above)
Comment (lundinc2) [reply libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:688]: Oops, moving this

File: projects/cypress/CY8CKIT_064S0S2_4343W/mtb/jitp/config_files/iot_pkcs11_config.h [hunk 5]
```diff
@@ -32,9 +32,21 @@
...
 #define _AWS_PKCS11_CONFIG_H_
 
+/**************************************************/
+/******* DO NOT CHANGE the following order ********/
+/**************************************************/
+
+/* Include logging header files and define logging macros in the following order:
+ * 1. Include the header file "logging_levels.h".
+ * 2. Define the LIBRARY_LOG_NAME and LIBRARY_LOG_LEVEL macros depending on
+ * the logging configuration for MQTT.
```
Comment (aggarw13) [inline projects/cypress/CY8CKIT_064S0S2_4343W/mtb/jitp/config_files/iot_pkcs11_config.h:42]: Copy-pasta fixes across all config files!
```suggestion
 * the logging configuration for PKCS#11.
```

File: projects/cypress/CY8CKIT_064S0S2_4343W/mtb/jitp/config_files/iot_pkcs11_config.h [hunk 6]
```diff
@@ -32,9 +32,21 @@
...
+ * 3. Include the header file "logging_stack.h", if logging is enabled for MQTT.
```
Comment (aggarw13) [inline projects/cypress/CY8CKIT_064S0S2_4343W/mtb/jitp/config_files/iot_pkcs11_config.h:43]: Copy-pasta fixes across all config files!
```suggestion
 * 2. Define the LIBRARY_LOG_NAME and LIBRARY_LOG_LEVEL macros depending on
 * the logging configuration for PKCS#11.
 * 3. Include the header file "logging_stack.h", if logging is enabled for PKCS#11.
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 7]
```diff
@@ -561,17 +561,17 @@ static CK_RV prvMbedTLS_Initialize( void )
 
         if( 0 != lMbedTLSResult )
         {
-            LogError( ( "Failed to seed the mbed TLS DRBG for the PKCS #11 module with error %s",
-                        mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ) ) );
-            LogError( ( "%s", mbedtlsLowLevelCodeOrDefault( lMbedTLSResult ) ) );
+            LogError( ( "Could not initialize PKCS #11. Failed to seed the mbed TLS DRBG for the PKCS #11 module with mbed TLS error = %s : %s.",
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:564]: Log message could be shortened as 
```suggestion
            LogError( ( "Could not initialize PKCS #11. Failed to seed the mbed TLS DRBG: mbed TLS error = %s : %s.",
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 8]
```diff
@@ -623,52 +624,61 @@ static CK_RV prvCertAttParse( CK_ATTRIBUTE * pxAttribute,
...
             if( pxAttribute->ulValueLen <= pkcs11configMAX_LABEL_LENGTH )
             {
                 *ppxLabel = pxAttribute;
             }
             else
             {
-                LogError( ( "Label length was not in the valid range. Please check the value of pkcs11configMAX_LABEL_LENGTH" ) );
+                LogError( ( "Failed parsing certificate template. Label length"       \
+                            "was not in the valid range. Found %lu and expected %lu." \
+                            "Consider updating pkcs11configMAX_LABEL_LENGTH.", pxAttribute->ulValueLen, pkcs11configMAX_LABEL_LENGTH ) );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:641]: Backslash escapes are only requires for macro definitions, and not for macro calls
```suggestion
               LogError( ( "Failed to parse certificate template. Label length "
                "not in valid range. Found %lu and expected %lu. "
                "Consider updating pkcs11configMAX_LABEL_LENGTH.",
                pxAttribute->ulValueLen, pkcs11configMAX_LABEL_LENGTH ) );
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 9]
```diff
@@ -623,52 +624,61 @@ static CK_RV prvCertAttParse( CK_ATTRIBUTE * pxAttribute,
...
 
             break;
 
         case ( CKA_CERTIFICATE_TYPE ):
+            LogDebug( ( "Found CKA_CERTIFICATE_TYPE." ) );
             ( void ) memcpy( pxCertificateType, pxAttribute->pValue, sizeof( CK_CERTIFICATE_TYPE ) );
 
             if( *pxCertificateType != CKC_X_509 )
             {
+                LogError( ( "Failed parsing certificate template. Certificate type was invalid. Expected CKC_X_509, but found %0X.", *pxCertificateType ) );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:653]: Prefix hex values with `0x`?
```suggestion
                LogError( ( "Failed parsing certificate template. Certificate type was invalid. Expected CKC_X_509, but found 0x%0X.", *pxCertificateType ) );
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 10]
```diff
@@ -623,52 +624,61 @@ static CK_RV prvCertAttParse( CK_ATTRIBUTE * pxAttribute,
...
 
         case ( CKA_CLASS ):
         case ( CKA_SUBJECT ):
+            LogDebug( ( "Received attribute type %0X and ignored it.", pxAttribute->type ) );
 
             /* Do nothing.  This was already parsed out of the template previously. */
             break;
 
         default:
+            LogError( ( "Failed parsing certificate template. Received an unknown template type with value %0X.", pxAttribute->type ) );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:681]:
```suggestion
            LogError( ( "Failed to parse certificate template. Received an unknown template type: 0x%0X.", pxAttribute->type ) );
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 11]
```diff
@@ -685,33 +695,36 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
     CK_RV xResult = CKR_OK;
     int32_t lMbedTLSResult = 0;
     CK_BBOOL xBool;
-    mbedtls_rsa_context * pxRsaContext = ( mbedtls_rsa_context * ) pxMbedContext->pk_ctx;
 
-    configASSERT( pxRsaContext != NULL );
+    configASSERT( ( mbedtls_rsa_context * ) pxMbedContext->pk_ctx != NULL );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:699]: What's the purpose of this change? I think the previous version is more readable for the NULL check with `configASSERT( pxRsaContext != NULL );`

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 12]
```diff
@@ -685,33 +695,36 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
...
         case ( CKA_TOKEN ):
+            LogDebug( ( "Found CKA_TOKEN or CKA_SIGN." ) );
             ( void ) memcpy( &xBool, pxAttribute->pValue, pxAttribute->ulValueLen );
 
             /* See explanation in prvCheckValidSessionAndModule for this exception. */
             /* coverity[misra_c_2012_rule_10_5_violation] */
             if( xBool != ( CK_BBOOL ) CK_TRUE )
             {
-                LogError( ( "Only RSA private keys with signing permissions supported." ) );
+                LogError( ( "Failed to parse RSA private key template. Expected signing permissions to be supported or for the object to be a token type." ) );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:720]: The second statement is not clear...Can it be changed to something like "Either token type object OR private keys with signing permissions are supported"?

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 13]
```diff
@@ -757,28 +774,30 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
...
             xResult = CKR_ATTRIBUTE_TYPE_INVALID;
             break;
     }
 
     if( lMbedTLSResult != 0 )
     {
-        LogError( ( "mbedTLS create private RSA key failed with error %s.",
-                    mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ) ) );
-        LogError( ( "%s.",
+        LogError( ( "Failed to parse RSA private key template. mbed TLS failed with mbed TLS error = %s : %s.",
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:799]:
```suggestion
        LogError( ( "Failed to parse RSA private key template. mbed TLS failed with error = %s : %s.",
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 14]
```diff
@@ -798,9 +817,9 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
         CK_BBOOL xBool = ( CK_BBOOL ) CK_FALSE;
         int32_t lMbedTLSResult = 0;
         CK_RV xResult = CKR_OK;
-        mbedtls_ecp_keypair * pxKeyPair = ( mbedtls_ecp_keypair * ) pxMbedContext->pk_ctx;
 
-        configASSERT( pxKeyPair != NULL );
+        configASSERT( ( mbedtls_ecp_keypair * ) pxMbedContext->pk_ctx != NULL );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:821]: Same comment about the assert expression change. I think the previous version is more readable

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 15]
```diff
@@ -811,7 +830,7 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
             if( xBool == ( CK_BBOOL ) CK_FALSE )
             {
                 xResult = CKR_ATTRIBUTE_VALUE_INVALID;
-                LogError( ( "Only EC private keys with signing privileges are supported." ) );
+                LogError( ( "Failed to parse EC private key. Expected signing privileges to be supported." ) );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:833]:
```suggestion
                LogError( ( "Failed to parse EC private key. Only private key with signing privileges are supported." ) );
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 16]
```diff
@@ -823,9 +842,8 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
             if( lMbedTLSResult != 0 )
             {
                 xResult = CKR_FUNCTION_FAILED;
-                LogError( ( "mbedTLS mpi read binary failed with error %s.",
-                            mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ) ) );
-                LogError( ( "%s.",
+                LogError( ( "Failed to parse EC private key. mbedTLS mpi read binary failed with mbed TLS error = %s : %s.",
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:845]:
```suggestion
                LogError( ( "Failed to parse EC private key. mbedTLS mpi read binary failed: mbed TLS error = %s : %s.",
```

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 17]
```diff
@@ -944,14 +965,17 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
                 }
                 else
                 {
-                    LogError( ( "Expected a public key for CKA_VERIFY or CKA_EC_POINT." ) );
+                    LogError( ( "Failed parsing EC key template. The key type"    \
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:968]: The backslashes can be removed in the logging calls. Also, spaces between the line broken words should be added

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c (hunk 11 above)
Comment (lundinc2) [reply libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:699]: I'll revert this

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 18]
```diff
@@ -3891,40 +4175,48 @@ CK_DECLARE_FUNCTION( CK_RV, C_VerifyInit )( CK_SESSION_HANDLE hSession,
...
+                        "take xVerifyMutex." ) );
             xResult = CKR_CANT_LOCK;
         }
     }
 
     /* Check that the mechanism and key type are compatible, supported. */
     if( xResult == CKR_OK )
     {
-        xKeyType = mbedtls_pk_get_type( &pxSession->xSignKey );
+        xKeyType = mbedtls_pk_get_type( &pxSession->xVerifyKey );
```
Comment (aggarw13) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:4187]: Is this change intentional?

Comment (lundinc2) [reply libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:4187]: Yes I stumbled on this bug when doing these logging updates.

File: libraries/c_sdk/standard/common/include/logging_stack.h [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+/* FreeRTOS Include. */
+#include "FreeRTOS.h"
+
+/* Standard Include. */
+#include <stdint.h>
+
+/* Metadata information to prepend to every log message. */
+#define LOG_METADATA_FORMAT    "[%s:%d] "
+#define LOG_METADATA_ARGS      __FILE__, __LINE__
```
Comment (aggarw13) [inline libraries/c_sdk/standard/common/include/logging_stack.h:42]: I recently updated the metadata on C-SDK to log the just the file name instead of the absolute file path. The same change can be brought here as well.
Reference: https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/development/demos/logging-stack/logging_stack.h#L39-L45

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 20]
```diff
@@ -1700,6 +1788,8 @@ CK_DECLARE_FUNCTION( CK_RV, C_GetMechanismInfo )( CK_SLOT_ID slotID,
     if( pInfo == NULL )
     {
         xResult = CKR_ARGUMENTS_BAD;
+        LogError( ( "Failed to get mechanism info. pInfo was NULL. Expected a "
+                    "pointer to a valid CK_MECHANISM_INFO struct." ) );
     }
```
Comment (gkwicker) [inline libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:1793]: Outside of the scope of this PR, but for readability I would move 1788-1793 to 1811 and change 1812 to an `else`.

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c (hunk 18 above)
Comment (gkwicker) [reply libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:4187]: Good catch – I would search this file for the same/similar error if you haven't already.

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c [hunk 21]
```diff
@@ -849,23 +881,22 @@ static CK_RV prvRsaKeyAttParse( const CK_ATTRIBUTE * pxAttribute,
...
         else
         {
-            lMbedReturn = mbedtls_ecp_point_read_binary( &pxKeyPair->grp,
-                                                         &pxKeyPair->Q,
-                                                         ( ( uint8_t * ) ( pxAttribute->pValue ) + 2U ),
-                                                         ( pxAttribute->ulValueLen - 2U ) );
+            lMbedTLSResult = mbedtls_ecp_point_read_binary( &pxKeyPair->grp,
+                                                            &pxKeyPair->Q,
+                                                            ( ( uint8_t * ) ( pxAttribute->pValue ) + 2U ),
+                                                            ( pxAttribute->ulValueLen - 2U ) );
```
[the remark this answers is not in the document]
Comment (lundinc2) [reply libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:892]: Thanks I will add a comment here, this strips some binary encoding data that mbed TLS does not like, off the top of my head I believe its indicating the object type and if it's compressed, I'll refresh my memory and add comments

File: libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c (hunk 20 above)
Comment (lundinc2) [reply libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c:1793]: Noted, I will update this

File: vendors/infineon/boards/xmc4800_plus_optiga_trust_x/aws_demos/config_files/iot_pkcs11_config.h [hunk 22]
```diff
@@ -36,25 +36,6 @@
 /******* DO NOT CHANGE the following order ********/
```
Comment (aggarw13) [inline vendors/infineon/boards/xmc4800_plus_optiga_trust_x/aws_demos/config_files/iot_pkcs11_config.h:36]: You may want to remove these comment sections as well
